Formula/infer.rb

```diff
@@ -29,33 +35,42 @@ def install
     # Apple's libstdc++ is too old to build LLVM
     ENV.libcxx if ENV.compiler == :clang
 
-    opamroot = buildpath/"opamroot"
-    opamroot.mkpath
+    ENV.prepend_path "PKG_CONFIG_PATH", Formula["sqlite"].opt_lib/"pkgconfig"
+
+    # opamroot = HOMEBREW_CACHE/"opam"
+    opamroot = libexec/"opam"
     ENV["OPAMROOT"] = opamroot
     ENV["OPAMYES"] = "1"
 
-    # do not attempt to use the clang in facebook-clang-plugins/ as it hasn't been built yet
-    ENV["INFER_CONFIGURE_OPTS"] = "--prefix=#{prefix} --without-fcp-clang"
+    # explicitly build the clang before infer's configure
+    ENV["INFER_CONFIGURE_OPTS"] = "--prefix=#{prefix} --with-fcp-clang"
 
     llvm_args = %w[
       -DLLVM_INCLUDE_DOCS=OFF
       -DLLVM_INSTALL_UTILS=OFF
-      -DLLVM_TARGETS_TO_BUILD=all
       -DLIBOMP_ARCH=x86_64
-      -DLLVM_BUILD_EXTERNAL_COMPILER_RT=ON
-      -DLLVM_BUILD_LLVM_DYLIB=ON
     ]
 
-    system "opam", "init", "--no-setup"
-    ocaml_version = File.read("build-infer.sh").match(/OCAML_VERSION_DEFAULT=\"([^\"]+)\"/)[1]
-    ocaml_version_number = ocaml_version.split("+", 2)[0]
-    inreplace "#{opamroot}/compilers/#{ocaml_version_number}/#{ocaml_version}/#{ocaml_version}.comp",
-      '["./configure"', '["./configure" "-no-graph"'
+    pathfix_lines = [
+      "export SDKROOT=#{MacOS.sdk_path}",
+      "eval $(opam env)"
+    ]
+
     # so that `infer --version` reports a release version number
-    inreplace "infer/src/base/Version.ml.in", "let is_release = is_yes \"@IS_RELEASE_TREE@\"", "let is_release = true"
+    inreplace "infer/src/base/Version.ml.in", "@IS_RELEASE_TREE@", "yes"
+
     inreplace "facebook-clang-plugins/clang/setup.sh", "CMAKE_ARGS=(", "CMAKE_ARGS=(\n  " + llvm_args.join("\n  ")
-    system "./build-infer.sh", "all", "--yes"
-    system "opam", "config", "exec", "--switch=infer-#{ocaml_version}", "--", "make", "install"
+
+    # setup opam (disable opam sandboxing because homebrew is sandboxed already)
+    inreplace "build-infer.sh", "--no-setup", "--no-setup --disable-sandboxing"
+    # prefer system bins for opam install
+    inreplace "build-infer.sh", "opam install", "PATH=/usr/bin:$PATH\n    opam install"
```

Comment:
Is this because you need system opam? Or because opam won't work with brewed stuff? This still seems very hacky. At a minimum a comment is needed as to why this is, but ideally this complexity should be reduced or at least managed by upstream.

Reply:
This workaround is because one of the opam dependencies, `mlgmpidl`, needs a system dependency to get through the configure phase. I think it's related to system clang rather than the homebrew shim, but I honestly have no clue what or why. This is only a problem inside the homebrew build environment; running this command from a normal mac shell works fine.